### PR TITLE
57 unscanned tag is constantly deleted

### PR DESCRIPTION
**Pull request by unglaublicherdude**
This PR does now a number of things.

1. It fixes both tag problems

     1. The Unscanned tag will not be created constantly, because the ScanService will not create it constantly anymore
     2. The TagService will now not just remove all tags from files it will tag, it will lookup the tags for a file and just create and delete what's needed

2. Adding unittest

I addes some tests run by phpunit as we already have in the vaas-sdk. You can take these tests as example, to add your own tests.

3. Adding some devcontainer features

The most important one is the docker-in-docker feature. This means you can now run everything within the devcontainer. So you start the devcontainer, run the ./install.sh script and it will run nextcloud within the devcontainer.  Meaning to go to the nextcloud page running in this container, you can just click on the browser-icon in the ports tab in your vscode.

image

5. overriding a dependency of webpack-vue-config as it has a vulnerability marked as high 

```json
  "overrides": {
    "@nextcloud/webpack-vue-config": {
      "webpack-dev-server": "5.0.4"
    }
  }
```

### CODE REVIEW

File: lib/Command/GetTagIdCommand.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		$this->logger = $logger;
+		$this->tagService = $tagService;
+	}
+
+	/**
+	 * @return void
+	 */
+	protected function configure() {
+		$this->setName('gdatavaas:get-tag-id');
+		$this->setDescription('tags all files without tag from this app as unscanned');
```
Comment (pstadermann) [inline lib/Command/GetTagIdCommand.php:30]: Description does not match function. Maybe "Gets the ID of a tag".

File: lib/Command/RemoveTagCommand.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		$this->logger = $logger;
+		$this->tagService = $tagService;
+	}
+
+	/**
+	 * @return void
+	 */
+	protected function configure() {
+		$this->setName('gdatavaas:remove-tag');
+		$this->setDescription('tags all files without tag from this app as unscanned');
```
Comment (pstadermann) [inline lib/Command/RemoveTagCommand.php:30]:
```suggestion
		$this->setDescription('deletes a tag');
```

File: lib/Command/RemoveTagCommand.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		$this->addArgument(self::TAG_NAME, InputArgument::REQUIRED, "Tag name want to delete.");
```
Comment (pstadermann) [inline lib/Command/RemoveTagCommand.php:31]:
```suggestion
		$this->addArgument(self::TAG_NAME, InputArgument::REQUIRED, "The name of the tag name you want to delete.");
```

File: lib/Service/TagService.php [hunk 4]
```diff
@@ -57,11 +57,24 @@ public function getTag(string $name, bool $create = true): ISystemTag {
...
 	 */
 	public function setTag(int $fileId, string $tagName): void {
-		$tag = $this->getTag($tagName);
+		$tag = $this->tagService->getTag($tagName, true, false);
+		$filesTagIds = $this->tagMapper->getTagIdsForObjects($fileId, 'files');
+
+		if (isset($filesTagIds[$fileId])) {
+			foreach ($filesTagIds[$fileId] as $tagId) {
+				if ($tagId != $tag->getId()) {
+					$this->tagMapper->unassignTags(strval($fileId), 'files', [$tagId]);
```
Comment (pstadermann) [inline lib/Service/TagService.php:66]: Doesn't this remove all tags, not only ours?

**Review by pstadermann — CHANGES_REQUESTED** [review]
Currently removes all other tags when setting. I thought we should only remove our tags.

File: lib/Command/RemoveTagCommand.php (hunk 2 above)
Comment (lennartdohmann) [reply lib/Command/RemoveTagCommand.php:30]: Same here

**Review by lennartdohmann — CHANGES_REQUESTED** [review]

**Review by doxthree — APPROVED** [review]

**Review by lennartdohmann — APPROVED** [review]

File: tests/unittests/TagServiceTest.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+		$tagMapper = $this->createMock(ISystemTagObjectMapper::class);
+		$tagMapper->method('getTagIdsForObjects')->willReturn([self::$OBJECT_ID_1 => [TagService::CLEAN, TagService::MALICIOUS]]);
+		$unassignTagsMatcher = $this->exactly(2);
+		$tagMapper->expects($unassignTagsMatcher)->method('unassignTags')->willReturnCallback(function ($objectId, $objectType, $tagIds) use ($unassignTagsMatcher) {
+			switch($unassignTagsMatcher->numberOfInvocations()) {
+				case 1:
+					$this->assertEquals($objectId, strval(self::$OBJECT_ID_1));
+					$this->assertEquals($objectType, 'files');
+					$this->assertEquals($tagIds, [TagService::CLEAN]);
```
Comment (lennartdohmann) [inline tests/unittests/TagServiceTest.php:37]: The assertions are misordered. The actual value should be second.
Screenshot 2024-07-11 093658

File: lib/Command/GetTagIdCommand.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		$this->setDescription('Gets the ID of a tag');
+		$this->addArgument(self::TAG_NAME, InputArgument::REQUIRED, "Tag name you want to get the id for.");
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output): int {
+		$logger = new ConsoleCommandLogger($this->logger, $output);
+		$tagName = $input->getArgument('tag-name');
+		try {
+			$tag = $this->tagService->getTag($tagName, false);
+		} catch (TagNotFoundException $e) {
```
Comment (lennartdohmann) [inline lib/Command/GetTagIdCommand.php:39]: I run into issues because this is not imported.

File: lib/Command/GetTagIdCommand.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace OCA\GDataVaas\Command;
+
+use OCA\GDataVaas\Logging\ConsoleCommandLogger;
+use OCA\GDataVaas\Service\TagService;
```
Comment (lennartdohmann) [inline lib/Command/GetTagIdCommand.php:6]:
```suggestion
use OCP\SystemTag\TagNotFoundException;
```

File: lib/Command/GetTagIdCommand.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	public function __construct(TagService $tagService, LoggerInterface $logger) {
+		parent::__construct();
+		$this->logger = $logger;
+		$this->tagService = $tagService;
+	}
+
+	/**
+	 * @return void
+	 */
+	protected function configure() {
```
Comment (lennartdohmann) [inline lib/Command/GetTagIdCommand.php:28]:
```suggestion
	protected function configure(): void {
```

**Comment by lennartdohmann** [comment]
@unglaublicherdude I opened [this PR](https://github.com/GDATASoftwareAG/nextcloud-gdata-antivirus/pull/76) into this with my suggestions.

**Review by pstadermann — APPROVED** [review]

**Pull request merged**